container-search/src/main/java/com/yahoo/search/ranking/GlobalPhaseHelper.java

```
@@ -0,0 +1,152 @@
+                String queryFeatureName = ref.simpleArgument().get();
+                // searchers are recommended to place query features here:
+                var feature = rankFeatures.getTensor(queryFeatureName);
+                if (feature.isPresent()) {
+                    result.add(new NameAndValue(needed, feature.get()));
+                } else {
+                    // but other ways of setting query features end up in the properties:
+                    var objList = rankProps.get(queryFeatureName);
+                    if (objList != null && objList.size() == 1 && objList.get(0) instanceof Tensor t) {
+                        result.add(new NameAndValue(needed, t));
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    void rerankHits(Query query, Result result, HitRescorer hitRescorer, int rerank) {
+        double worstRerankedScore = Double.MAX_VALUE;
+        double worstRerankedOldScore = Double.MAX_VALUE;
+        // TODO consider doing recursive iteration instead of deepIterator
+        for (var iterator = result.hits().deepIterator(); iterator.hasNext();) {
+            Hit hit = iterator.next();
+            if (hit.isMeta() || hit instanceof HitGroup) {
+                continue;
+            }
+            // what about hits inside grouping results?
+            if (rerank > 0) {
+                double oldScore = hit.getRelevance().getScore();
+                boolean didRerank = hitRescorer.rescoreHit(hit);
+                if (didRerank) {
+                    double newScore = hit.getRelevance().getScore();
+                    if (oldScore < worstRerankedOldScore) worstRerankedOldScore = oldScore;
+                    if (newScore < worstRerankedScore) worstRerankedScore = newScore;
+                    --rerank;
+                } else {
+                    // failed to rescore this hit, what should we do?
+                    hit.setRelevance(-Double.MAX_VALUE);
```

When can this happen?

---

this should only fail if something is missing from match-features.  after looking at how score adjustment happens in C++ after second-phase I am implementing the same algorithm here.